scrapy/extensions/robotstxtparser.py

```diff
@@ -0,0 +1,137 @@
+import sys
+import logging
+
+from scrapy.utils.python import to_native_str, to_unicode
+
+logger = logging.getLogger(__name__)
+
+class PythonRobotParser():
+    def __init__(self, content, spider):
+        from six.moves.urllib_robotparser import RobotFileParser
+        self.spider = spider
+        try:
+            content = to_native_str(content)
+        except UnicodeDecodeError:
+            # If we found garbage or robots.txt in an encoding other than UTF-8, disregard it.
+            # Switch to 'allow all' state.
+            logger.warning("Failure while parsing robots.txt using %(parser)s." 
+                "File either contains garbage or is in an encoding other than UTF-8, treating it as an empty file.",
+                {'parser': "RobotFileParser"},
+                exc_info=sys.exc_info(),
+                extra={'spider': self.spider})
+            content = ''
+        self.rp = RobotFileParser()
+        self.rp.parse(content.splitlines())
+
+    @classmethod
+    def from_crawler(cls, crawler, content):
+        spider = None if not crawler else crawler.spider
+        o = cls(content, spider)
+        return o
+
+    def allowed(self, url, user_agent):
+        user_agent = to_native_str(user_agent)
+        url = to_native_str(url)
+        return self.rp.can_fetch(user_agent, url)
+
+    def sitemaps(self):
+        """RobotFileParser does not support Sitemaps directive."""
+        logger.warning("Failure while retrieving sitemaps using %(parser)s." 
+            "%(parser)s does not support Sitemaps directive.",
+            {'parser': "RobotFileParser"},
+            exc_info=sys.exc_info(),
+            extra={'spider': self.spider})
+        return
+        yield
+
+    def crawl_delay(self, user_agent):
+        """RobotFileParser does not support Crawl-delay directive for Python version < 3.6 ."""
+        user_agent = to_native_str(user_agent)
+        if hasattr(self.rp, 'crawl_delay'):
+            delay = self.rp.crawl_delay(user_agent)
+            return None if delay is None else float(delay)
+        return None
+
+    def preferred_host(self):
+        """RobotFileParser does not support Host directive."""
+        logger.warning("Failure while retrieving preferred host using %(parser)s." 
+            "%(parser)s does not support Host directive.",
+            {'parser': "RobotFileParser"},
+            exc_info=sys.exc_info(),
+            extra={'spider': self.spider})
+        return None
+
```

Comment:
nitpick: according to pep8, there should be two blank lines between class definitions